repository-s3/src/main/java/io/aiven/elasticsearch/repositories/s3/S3RepositoryStorageIOProvider.java

```diff
@@ -0,0 +1,271 @@
+            return Permissions.doPrivileged(() -> {
+                final S3ObjectInputStream objectContent;
+                try {
+                    objectContent = client.getObject(bucketName, blobName).getObjectContent();
+                } catch (final AmazonClientException e) {
+                    throw new IOException("Couldn't read blob " + blobName, e);
+                }
+                return cryptoIOProvider.decryptAndDecompress(objectContent);
+            });
+        }
+
+        @Override
+        public void write(final String blobName,
+                          final InputStream inputStream,
+                          final long blobSize,
+                          final boolean failIfAlreadyExists) throws IOException {
+            try {
+                Permissions.doPrivileged(() ->
+                        cryptoIOProvider.compressAndEncrypt(
+                                inputStream,
+                                new S3OutputStream(bucketName, blobName, partSize, client)
+                        )
+                );
+            } catch (final AmazonClientException e) {
+                throw new IOException("Couldn't upload blob with name: " + blobName, e);
+            }
+        }
+
+        @Override
+        public Tuple<Integer, Long> deleteDirectories(final String path) throws IOException {
+            final List<Tuple<String, Long>> listOfFiles;
+            try {
+                //we need only prefix here since it is impossible to distinguish between "files" and "directories"
+                //in AWS SDK e.g. paths like:
+                // - /aaaa/bbb/ccc/file
+                // - /aaa/bbb/file
+                // with request which contains delimiter and without returns diff result, and ES API doesn't use
+                // recursive calls to get full list of files by path/sub-path
+                listOfFiles = Permissions.doPrivileged(() ->
+                        list(new ListObjectsV2Request().withBucketName(bucketName).withPrefix(path)))
+                        .stream()
+                        .flatMap(l -> l.getObjectSummaries().stream())
+                        .map(o -> Tuple.tuple(o.getKey(), o.getSize()))
+                        .collect(Collectors.toList());
+            } catch (final AmazonClientException e) {
+                throw new IOException("Couldn't get list of files for path " + path, e);
+            }
+
+            deleteFiles(listOfFiles.stream()
+                            .map(Tuple::v1)
+                            .collect(Collectors.toList()),
+                    true);
+
+            final var removedBytes = listOfFiles.stream().map(Tuple::v2).reduce(0L, Long::sum);
+            return Tuple.tuple(listOfFiles.size(), removedBytes);
+        }
+
+        @Override
+        public void deleteFiles(final List<String> blobNames,
+                                final boolean ignoreIfNotExists) throws IOException {
```

Comment:
`ignoreIfNotExists` is not used

Reply:
yes S3 will raise `MultiObjectDeleteException` which will contain problematic files, which will be logged. Compare to GCS it doesn't support special headers for ignoring such files.

Reply:
Ah, I got it.

> S3 will raise `MultiObjectDeleteException` which will contain problematic files

I have some doubts about this part as `withQuiet` is used. Could you check it really happens?